### ISSUE TITLE
Unclear execution of setup / test case / teardown

### ISSUE THREAD

**Issue by lukasradek** [psiho/jsbench-me#54]
I was getting some inconsistent results and find out the following.

My **test case** is sometimes but regularly executed multiple times after single **setup javascript** like this.
1x Setup JS
1x Test Case
1x Setup JS
1x Test Case
1x Setup JS
Many times Test Case

Since the test cannot be setup again, it leads to unconsistent results.

Example https://jsbench.me/v9kzlav8l6/1

If you open the console, you will see logs:
"Setup JS" - self explanatory
"Skipped **number**" - result of first test case, this is getting repeated

**Comment by psiho** [comment]
Cleaning up issues and ran into this old one. Probably too late to reply but might be of use to somebody else maybe...

JSBench.me is just a GUI to Benchmark.js, so to best understand how to setup tests ... Benchmark.js docs are the best place I guess. As far as I undestand it (and tbh I use it as a black box) this is by design. Setup code is run once before each LOOP, and we have many loops and many test runs inside each loop (depending on sped of each execution). That's why you get single setup run, followed by many code runs, and then repeated many times.

The reason you get a few single setup -> test -> setup -> test runs at the start is (I guess!) initialization or "warmup" procedure of Benchmark.js

Just take a look at https://benchmarkjs.com/docs/#prototype_setup and you can see compiled code which clearly shows what is run when.

More interesting read to find out about JS benchmarking in general: https://mathiasbynens.be/notes/javascript-benchmarking

**Comment by psiho** [comment]
It is a common misconception how the tests are run by Benchmark. Sometimes terminology is also in the way of understanding it. So, it depends what you mean by "loop". Because ther are two llops. One loop runs test X time and make it a "sample". Onother loop runs several cycles. SetupJS is executed before each samle cycle, NOT every test cycle!

Similar thread with one of the maintainers has it explained: https://github.com/bestiejs/benchmark.js/issues/183#issuecomment-271967126

But my take on it is, for a bechmark suite with a single test in it, in JSBenchme:
`
- SetupHTML (inline scripts)
    <samples loop X times>
        - Setup JS
            <start timer>
                <samle loop X times, so loop does not last longer than Y seconds>
                     - run test function once
                </end sample loop>
             </end timer>
         - Teardown JS
     </end samples loop>
`

Now, a lot of people ask "Why? Why can't I run SetupJS before each single test execution instead of before each sample loop?". Answer is complex, but bottom line this is to make test results better. IN the thread above there is a link to explanation: https://calendar.perfplanet.com/2010/bulletproof-javascript-benchmarks/

**Comment by lukasradek** [comment]
Alright 🙂, to be honest it is not really an issue for me since I don't use it anymore. So let's finish it here.

**Comment by psiho** [comment]
Well, it's really a good topic for any javascript benchmarking, which is far more complex than it looks.  So, this thread will be useful to others. Thanks for your time.

**Comment by lukasradek** [comment]
Likewise.

**Comment by lukasradek** [comment]
The exact test is right in my OP 😉. I was benching sorting methods and needed to setup unsorted array of numbers.

**Comment by psiho** [comment]
Oh man sorry, I'm juggling tickets, coding, doing lunch, all at the same time. Lost my head :)

From the top of my head, in SetupJS, I'd write a code that generates array of 1000 arrays of random numbers. I guess 1000 is enough and that there will not be more loops, but add up if it is not enough.

Then, you can use not so well documented feature, but... inside test loop, you have `items.count` available which counts number of executions within single cycle. So, within your test code, use `randoms[this.count]` instead of `randoms`.

Something along those lines anway. You get the point.

**Comment by lukasradek** [comment]
Yeah... that is why I was asking for setup before each iteration 😁
